lib/app/modules/smart/token_refresh_sequence.rb

```diff
@@ -28,133 +31,150 @@ class TokenRefreshSequence < SequenceBase
+        token_response = perform_refresh_request(INVALID_CLIENT_ID, @instance.refresh_token)
+        assert_response_bad_or_unauthorized token_response
+      end
 
-        assert @token_response_body.key?('access_token'), 'Token response did not contain access_token as required'
+      def validate_and_save_refresh_response(token_response)
+        assert_response_ok(token_response)
+        assert_valid_json(token_response.body)
+        token_response_body = JSON.parse(token_response.body)
+
+        assert token_response_body.key?('access_token'), 'Token response did not contain access_token as required'
 
         token_retrieved_at = DateTime.now
 
         @instance.resource_references.each(&:destroy)
-        @instance.resource_references << Inferno::Models::ResourceReference.new(resource_type: 'Patient', resource_id: @token_response_body['patient']) if @token_response_body.key?('patient')
+        @instance.resource_references << Inferno::Models::ResourceReference.new(resource_type: 'Patient', resource_id: token_response_body['patient']) if token_response_body.key?('patient')
 
         @instance.save!
 
-        @instance.update(token: @token_response_body['access_token'], token_retrieved_at: token_retrieved_at)
+        @instance.update(token: token_response_body['access_token'], token_retrieved_at: token_retrieved_at)
 
         ['token_type', 'scope'].each do |key|
-          assert @token_response_body.key?(key), "Token response did not contain #{key} as required"
+          assert token_response_body.key?(key), "Token response did not contain #{key} as required"
         end
 
         # case insentitive per https://tools.ietf.org/html/rfc6749#section-5.1
-        assert @token_response_body['token_type'].casecmp('bearer').zero?, 'Token type must be Bearer.'
+        assert token_response_body['token_type'].casecmp('bearer').zero?, 'Token type must be Bearer.'
 
         expected_scopes = @instance.scopes.split(' ')
-        actual_scopes = @token_response_body['scope'].split(' ')
+        actual_scopes = token_response_body['scope'].split(' ')
 
         warning do
           missing_scopes = (expected_scopes - actual_scopes)
           assert missing_scopes.empty?, "Token exchange response did not include expected scopes: #{missing_scopes}"
 
-          assert @token_response_body.key?('patient'), 'No patient id provided in token exchange.'
+          assert token_response_body.key?('patient'), 'No patient id provided in token exchange.'
         end
 
-        scopes = @token_response_body['scope'] || @instance.scopes
+        scopes = token_response_body['scope'] || @instance.scopes
 
         @instance.save!
         @instance.update(scopes: scopes)
 
-        if @token_response_body.key?('id_token')
+        if token_response_body.key?('id_token')
           @instance.save!
-          @instance.update(id_token: @token_response_body['id_token'])
+          @instance.update(id_token: token_response_body['id_token'])
```

Comment:
There shouldn't be, but DataMapper is fiddly about synchronized cached data and I'm concerned that removing this will cause bad behavior.  I would like to factor out DataMapper references in our test dsl and just think about it in terms of a state/context object that the tests can interact with, which is written so we don't have to ever do the 'save!' thing.